### ISSUE TITLE
Autonomous research campaigns: bounded agent-driven experimentation and improvement

### ISSUE THREAD

**Issue by mfittko** [mfittko/market-signals#231]
## Outcome

Implement a first-class **research campaign** that autonomously proposes, implements, evaluates and revises experimental candidates within a human-approved scope and budget.

**Approve the campaign once, not every experiment.** The deliverable is a reproducible candidate for separate qualification, or an evidence-backed negative/insufficient-evidence result. The researcher cannot change the authoritative evaluator, expand its permissions, or replace an operational trading deployment.

This is a focused implementation issue connecting #229 and #230, not another architecture epic or a commitment to a general autonomous strategy inventor.

| Owner | Responsibility |
|---|---|
| **#229 — runtime** | Supervised workers, capability profiles, durable runs/attempts, job dispatch, cancellation, budgets and restart-safe resumption |
| **#230 — research/evaluation** | Point-in-time datasets, `ExperimentSpec`, `ResearchRunner`, portfolio backtesting, forecast/outcome records, metrics and deployment qualification |
| **This issue — campaigns** | Search policy, candidate lineage, iteration, bounded research memory, stopping rules and submission of frozen candidates to qualification |

Dependencies are **specific capabilities**, not completion of both epics. Reuse their contracts and services; do not build another scheduler, backtester, registry or deployment authority.

## First vertical slice: E1, learn when to abstain

Automate #230's E1 on one selected instrument/feed and one defined prediction horizon, using the E0 deterministic baseline as the reference.

Keep baseline direction, position sizing, exits, opportunity stream, accounting and costs fixed. Allow candidate changes only to a declared entry-filter module: approved backward-looking features/transformations, regularized logistic-regression settings and the accept/skip threshold. Start with one model family; tree models and semantic evidence are extensions, not MVP dependencies.

Example approved objective:

> Determine whether an entry filter improves the fixed baseline after costs, subject to predeclared drawdown, coverage and evidence requirements. Explore only the registered feature and parameter space. Preserve unsuccessful trials and return a candidate dossier or an explicit negative result.

Before execution, choose the actual instrument, immutable dataset, target definition, split windows, objective/constraints and finite resource budgets. Missing mandatory values make a campaign invalid; do not silently invent trading thresholds or unlimited budgets.

**MVP boundary:** one researcher, one candidate job at a time, trusted numerical evaluation and an inspectable CLI/API. No agent swarm, new dashboard, Jev integration, operational paper trading or live account is required to demonstrate this loop.

## 1. Campaign contract and records

A versioned `CampaignSpec` wraps/references #230's `ExperimentSpec` rather than duplicating its methodology.

| Field group | Required content |
|---|---|
| Scope | Hypothesis, falsification condition, baseline, instrument/feed, target/horizon, allowed model families/features/parameter ranges and editable artifact paths |
| Evidence protocol | Dataset/split IDs, data-access policy, evaluator and cost-model versions, metrics, constraints, selection rule and protected qualification policy |
| Search policy | Initial candidate, permitted mutations, comparison/retention rule, repeat/seed policy and stopping conditions |
| Resources | Maximum distinct trials, technical retries, elapsed duration, compute, provider requests/tokens/spend, artifact volume and concurrency |
| Authority | Approver, approved spec hash, permitted tools/network destinations and any separately authorized forecast-only shadow scope |
| Reproducibility | Code/environment hashes, model/prompt versions, dependency lock/image, seeds, input/output schemas and artifact retention |

Changing the objective, dataset protocol, allowed search space, evaluator or authority creates a new approved campaign revision. Pausing/resuming does not reset budgets or evidence-access history.

Add campaign-specific records only where existing records are insufficient:

- **`ResearchCampaign`:** approved spec, lifecycle, aggregate budgets, current cursor and terminal reason.
- **`CandidateVersion`:** immutable parent, hypothesis, exact code/config diff, artifact hash and provenance. A retained development candidate is not a deployed champion.
- **`ResearchFinding`:** bounded claim, supporting/refuting run IDs, applicable instrument/windows/cost assumptions, evidence tier, creation time and validity/retraction status.
- **`QualificationSubmission`:** frozen candidate hash, evaluation protocol, exposure/submission record and resulting dossier.

Reference existing `ExperimentRun`, worker attempts, evaluation artifacts and forecast/outcome records. The service generates identifiers and records authoritative measurements; candidate code cannot declare itself successful.

## 2. Durable improvement loop

```mermaid
flowchart TD
    A[Approve campaign scope and budget] --> B[Load compact brief and eligible findings]
    B --> C[Propose hypothesis and candidate diff]
    C --> D[Validate allowed change and reserve budget]
    D --> E[Build and test in isolated workspace]
    E --> F[Trusted chronological development evaluation]
    F --> G[Record artifacts, measurements and failures]
    G --> H[Researcher proposes retain, revise or reject]
    H --> I{Controller checks policy and stopping rules}
    I -->|Continue within budget| B
    I -->|No candidate or insufficient evidence| N[Close campaign with evidence-backed report]
    I -->|Submit frozen candidate| Q[Protected qualification service]
    Q --> R[Candidate dossier and separately authorized next stage]
```

The researcher selects the next hypothesis and explains its expected effect. The controller validates its scope, the runner measures results, and the controller applies the registered retention/stopping rule. An LLM's favorable assessment never substitutes for numerical results or control checks.

Each iteration must preserve the candidate and its outcome before proceeding. Record invalid proposals, build failures, data failures, timeouts, rejected candidates and inconclusive comparisons, not just successful runs. Repairing candidate code creates a new version; an unchanged technical retry remains an attempt of the same run.

Use a campaign lifecycle such as `draft -> approved -> running <-> waiting_for_job | waiting_for_data -> completed`, with explicit `paused`, `cancelled`, `expired` and `failed` paths. Completion reasons include `candidate_submitted`, `no_qualified_candidate`, `insufficient_evidence` and `budget_exhausted`; a technical failure is not proof that a hypothesis failed.

Long training/evaluation jobs persist a job ID and release the agent worker. Resume from completion events or durable deadlines using #229's machinery. Deduplicate completion handling and fence superseded attempts. A late result may remain auditable but must not resurrect a cancelled campaign or become its authoritative next action.

Stop at the first applicable hard limit, a registered plateau rule, invalid evaluation/data, or operator cancellation. Do not implement “run until profitable.” A plateau is a search stopping decision, not statistical proof of no edge.

## 3. Research execution is a separate capability profile

Keep #229's execution-eligible trading agents restricted to approved evidence tools and typed intents. Do not grant them coding-agent privileges.

For this issue, introduce an explicit **research-worker profile** with candidate editing and bounded commands inside an ephemeral sandbox. Enforce the boundary outside prompts: scoped mounts and credentials, resource limits, denied host/operational access, controlled network egress and a trusted job gateway. No broker keys, production database, host shell, container socket or automatic writes to repository main.

Separate three authorities:

| Role | Can access/change | Cannot access/change |
|---|---|---|
| Researcher and candidate training | Approved development data, candidate workspace, permitted dependency environment and development feedback | Protected qualification data/results, authoritative evaluator, operating policy or deployments |
| Candidate inference during evaluation | Frozen candidate plus as-of inputs exposed by the runner | Future observations, evaluation labels, current news endpoints or later research memory |
| Trusted evaluator/controller | Required labels and account simulation, artifact/budget state, metric calculation and qualification controls | Accept candidate-reported scores as authoritative or grant self-promotion |

A read-only evaluator file in the same unrestricted process is insufficient. Run candidate code out of process; validate its predictions/actions and calculate metrics in the trusted service. Do not expose the full evaluation period or mutable label state to an inference subprocess. Imported candidate artifacts are untrusted code, including serialized-model loading.

The MVP can use typed edit/build/train/evaluate tools rather than a general shell. A bounded sandbox shell is an implementation option only after isolation tests pass. Dependencies are prebuilt/pinned or changed through a separately controlled build step. External LLM access, when needed for the researcher, goes through a budgeted gateway; it does not give historical inference arbitrary network access.

## 4. Independent evaluation and protected evidence

Reuse #230's chronological train/calibration/development protocol, point-in-time rules, portfolio accounting and cost/latency assumptions. Preprocessing and fitting must respect each permitted window. Development diagnostics may guide revisions; label-overlap controls still apply.

Freeze the authoritative metrics, baseline and comparison rule for the campaign. Evaluate net account results, drawdown, coverage/exposure, target-appropriate forecast quality, uncertainty, latency and total research cost. Prevent an apparent improvement obtained only by removing almost all opportunities. Do not collapse validity, forecast quality and economic value into an agent-written score.

Adaptive feedback itself can overfit reused evaluation data; ordinary holdout reuse is not automatically safe merely because inputs are chronological. See [Generalization in Adaptive Data Analysis and Holdout Reuse](https://arxiv.org/abs/1506.02629v2). This implementation does not claim a formal adaptive-testing guarantee.

**Qualification is outside the development loop:**

- The MVP permits at most one frozen finalist per campaign against the protected interval. The trusted service enforces submission policy; the research worker cannot query qualification opportunistically.
- Track protected-data access across related campaigns, candidates, researchers and shared memories. Starting a new campaign ID does not make an exposed interval untouched again.
- Keep qualification feedback out of continuing development. Once it influences a revision, record that exposure and require an appropriate new qualification protocol/future interval.
- Repeatedly checking forward results is also a selection process. Predefine review/maturation windows; do not repeatedly promote on the first favorable partial result.
- Historical calls to a newer LLM remain retrospective experiments. Prospective forecasts/outcomes are required by #230 for AI-influenced deployment qualification.

If an evaluator/data defect is found, halt affected comparisons, record the defect and mark impacted artifacts invalid/superseded. Fixes belong to the trusted implementation/review workflow. Re-run comparable candidates under a new version before comparing them; never let a candidate improve its score by modifying the exam.

## 5. Evidence-backed memory, not an endlessly growing conversation

Reconstruct each iteration from the approved brief, current candidate, compact experiment index and a bounded set of relevant findings. Preserve full transcripts/artifacts separately for audit, not as mandatory prompt history.

A finding must be scoped, for example: “Feature group X did not improve these candidate versions on these development windows under cost model Y; runs A/B/C.” Do not turn that into an unconditional instruction to avoid X forever.

Separate proposed explanations, measured development findings and reviewed conclusions. Retrieval must enforce evidence tier, temporal scope and campaign permissions before content reaches the agent. Store links to both supporting and contradicting results; permit retraction/supersession without deleting history.

Research memory may inform later research. It must not enter an earlier simulated trading decision or overwrite approved trading rules. Persist the context-manifest IDs actually supplied to each iteration and each evaluated decision so leakage and repeated misconceptions can be investigated.

## 6. Budgets, reproducibility and comparison discipline

Budget enforcement belongs to the controller/gateway, not the model. Reserve capacity/spend before dispatch and reconcile actual usage afterward. Track committed, outstanding and uncertain external spend; cancellation or timeout must not automatically refund a possibly executed provider call. Retries consume resources and cannot bypass campaign caps.

Hash dataset/splits, candidate, evaluator, environment, parameters, seed, model/prompt/tool versions and recorded provider outputs. Cache only under compatible keys and permissions. A repeated seed or model call is another observation, not permission to select the best-looking answer. Replaying a captured model response is distinguishable from making a fresh stochastic call.

Make validity failures visible and report numerical tolerances where deterministic byte-for-byte comparison is not appropriate. Reproducibility includes configurations and artifact lineage, not a promise that all fresh LLM outputs are identical.

**Measure the research agent as well as its candidates:** run a simple seeded random/parameter-search control over the same permitted search space, evaluator and comparable trial/compute budget. Include the agent's extra inference cost. Report best valid development candidate, robustness, failed trials and cost; do not claim the search method is superior from one lucky campaign. Any comparative final qualification must be prespecified so it does not create extra untracked holdout queries.

## 7. Minimal tool and operator surface

Expose versioned actions such as `get_campaign_brief`, `list_candidates`, `read_development_result`, `propose_candidate`, `submit_experiment`, `get_job_status`, `record_finding` and `request_qualification`. These are logical contracts, not a requirement for separate services. The gateway supplies/enforces campaign identity and capabilities; arguments cannot point to another campaign's dataset or evaluator.

The operator needs create/approve/start/pause/resume/cancel/inspect/report actions. Show the current hypothesis and candidate diff, budget usage, job state, comparison results, findings and exact stop reason. Start with CLI/API and Markdown/JSON reports; integrate into #229's run console later.

The campaign dossier contains the frozen candidate/configuration, full lineage/trial index, baseline comparisons, stress/coverage results, uncertainty and evidence limitations, resource use, failed hypotheses and qualification/prospective next steps.

No approval is required for each permitted development iteration. Automatic admission to isolated **forecast-only shadow** can be a later capability under an explicit research policy and quota. That is not approval to operate the shared paper portfolio. Activation of an operational strategy, any risk-policy change and all eventual P3 authority remain under #230's separate approval process.

## Delivery slices and capability dependencies

| Slice | Deliverable | Depends on |
|---|---|---|
| **C0 — contracts and fixture campaign** | Versioned spec, state machine, lineage and a deterministic fake researcher/runner exercising continue, reject and stop | #229 durable run/artifact contracts; #230 experiment/result schemas |
| **C1 — real E1 loop** | One sandboxed researcher, candidate editing, trusted E0/E1 evaluation, finite budgets and persisted development feedback | #229 research-worker/job capabilities; #230 one qualified dataset, baseline and runner |
| **C2 — evaluation isolation and recovery** | Qualification boundary, evidence-access ledger, restart/cancel/budget tests, memory isolation and tamper tests | C1 plus #230 qualification interface |
| **C3 — end-to-end evidence** | Unattended bounded campaign, simple-search comparison, replayable dossier and explicit outcome | C2; no full migration/dashboard/live broker required |

C0 demonstrates orchestration only; it must not be presented as research evidence. C1 requires the actual evaluation/data capabilities to be trustworthy before interpreting results.

Later extensions reuse this campaign abstraction for #230 E2 semantic-feature/Jev question refinement, E3 bounded confirmation workflows and E4 strategy selection. Prompt/tool-workflow optimization should be a distinct campaign with its own frozen objective, not simultaneous mutation of the controller and the candidate it evaluates.

## Acceptance criteria

- [ ] One approved E1 campaign completes without per-experiment operator input and produces a candidate dossier or explicit negative/insufficient-evidence outcome.
- [ ] Every candidate, parent/diff, selection attempt, technical retry, result and stop reason remains inspectable, including failures and rejected proposals.
- [ ] A restart during a job, duplicate completion, stale worker and cancellation cannot duplicate authoritative effects or resume cancelled work; budgets are preserved.
- [ ] Attempts to change evaluator/cost assumptions, exceed editable scope, read protected data, access future inference inputs, obtain operational credentials or promote a deployment are blocked and tested.
- [ ] Qualification exposure is tracked beyond a single campaign ID; protected results cannot leak through findings, caches, artifacts or agent context.
- [ ] The controller enforces trial/time/compute/provider budgets, including concurrent reservations, retries and uncertain external spend.
- [ ] Known leakage/evaluator-defect fixtures invalidate affected results rather than producing a better score.
- [ ] Retention/comparison uses trusted measurements, fixed rules and visible coverage/uncertainty; all-abstain and sparse-sample candidates cannot masquerade as demonstrated improvements.
- [ ] Captured-output replay reproduces deterministic artifacts within declared tolerances and attributes fresh stochastic calls separately.
- [ ] Research memory is bounded, evidence-linked and scope-aware; later findings cannot enter an earlier evaluated decision.
- [ ] A simple-search control and the full campaign resource costs are reported without assuming agent-led search is superior.
- [ ] Neither an operational deployment nor #229's trading-worker permissions changes merely because the campaign finds a promising candidate.

## Supporting patterns and non-goals

[GEPA](https://arxiv.org/abs/2507.19457v2) provides a relevant pattern for using execution feedback and reflection to propose and test prompt revisions. It is an optional later optimizer to assess, not an MVP dependency or evidence of trading performance. The adaptive-data-analysis reference above motivates the protected evaluation boundary; the detailed implementation here is a proposed engineering design, not a claim of equivalent statistical guarantees.

Non-goals: a new runtime or backtester; autonomous modification of evaluators/controllers; unrestricted strategy invention; model-weight fine-tuning/RL; multi-agent swarms; automatic operational deployment; live orders; mandatory Jev/GEPA adoption; and guaranteeing that search discovers an economic edge.

**Success means the system conducts a controlled, auditable experiment cycle—not that it is required to find a profitable strategy.**
